### PR TITLE
adds support for invoking closures

### LINKED ISSUE #3848: Support for closures or at least `Closure::__invoke`
Calling a closure is not supported in twig right now, any variable passed as `\Closure` cannot be called directly as Twig looks up the function in global functions (and extensions), the alternative is to call it via `__invoke` method on closure.

https://3v4l.org/EBR2Y (`__invoke` method does exist on Closures and work the same way as `$closure()` works).

Unfortunately this doesn't work in twig as `twig_get_attribute` uses `get_class_methods` to determine available methods on an object and fallbacks to `__call` if it exists and php  being php `get_class_methods($closure)` doesn't return `__invoke` https://3v4l.org/VKjAF :man_shrugging:  even though it exists and [is documented](https://www.php.net/manual/en/class.closure.php).

I think this can be fixed by checking if `$object` is instance of \Closure and then allow calling `__invoke` or is there a better way ?

I am happy to contribute a PR if it makes sense.

### PR DESCRIPTION
**Pull request by faizanakram99**
- fixes #3848

### CODE REVIEW

**Comment by faizanakram99** [comment]
@fabpot errors are unrelated

File: tests/TemplateTest.php [hunk 1]
```diff
@@ -394,6 +394,13 @@ public static function getGetAttributeTests()
             [true, ['foo' => 'bar'], $arrayAccess, 'vars', [], $anyType],
         ]);
 
+        // test for Closure::__invoke()
+        $closure = function (): string {
+            return 'closure called';
+        };
```
Comment (fabpot) [inline tests/TemplateTest.php]:
```suggestion
        // test for Closure::__invoke()
        $closure = fn (): string => 'closure called';
```

**Review by fabpot — CHANGES_REQUESTED** [review]

File: tests/TemplateTest.php [hunk 2]
```diff
@@ -394,6 +394,11 @@ public static function getGetAttributeTests()
             [true, ['foo' => 'bar'], $arrayAccess, 'vars', [], $anyType],
         ]);
 
+        // test for Closure::__invoke()
+        $closure = fn (): string => 'closure called';
+
+        $tests[] = [true, 'closure called', $closure, '__invoke', [], $anyType];
+
```
Comment (fabpot) [inline tests/TemplateTest.php]: Now that I see it, maybe we can even inline everything!

```suggestion
        $tests[] = [true, 'closure called', fn (): string => 'closure called', '__invoke', [], $anyType];
```

**Comment by fabpot** [comment]
Last comment and I think we're good: can you add an item in the CHANGELOG file?

**Review by nicolas-grekas — APPROVED** [review]
LGTM
I'd suggest opening an issue on php-src also, get_class_methods() should return __invoke to me (and other non static methods BTW).

**Comment by faizanakram99** [comment]
> Last comment and I think we're good: can you add an item in the CHANGELOG file?

Done

**Comment by faizanakram99** [comment]
> LGTM I'd suggest opening an issue on php-src also, get_class_methods() should return __invoke to me (and other non static methods BTW).

Done https://github.com/php/php-src/issues/17126

**Review by lowwa132 — COMMENTED** [review]
Thank you for the PR @faizanakram99

File: tests/TemplateTest.php [hunk 3]
```diff
@@ -394,6 +394,9 @@ public static function getGetAttributeTests()
             [true, ['foo' => 'bar'], $arrayAccess, 'vars', [], $anyType],
         ]);
 
+        // test for Closure::__invoke()
+        $tests[] = [true, 'closure called', fn (): string => 'closure called', '__invoke', [], $anyType];
```
Comment (fabpot) [inline tests/TemplateTest.php:398]: Let's add another test:

```
$tests[] = [true, 'closure called', fn (): string => 'closure called', '__invoke', [], $methodType];
```

**Review by fabpot — CHANGES_REQUESTED** [review]
For PHP 8.0, we need to check that line 1742, we not calling `isset($object->$item)` for Closure as it does not work.

**Comment by fabpot** [comment]
Thank you @faizanakram99.

**Pull request merged**